File: convert_model.py
#!/usr/bin/env python3
"""
╔══════════════════════════════════════════════════════════╗
║   AgroMind — convert_model.py                           ║
║   Converts AgroMind.keras → TF.js format                ║
║   Run this ONCE before opening the web app.             ║
╚══════════════════════════════════════════════════════════╝

STEP 1 — Install dependencies (run once):
    pip install tensorflowjs tensorflow

STEP 2 — Run this script:
    python convert_model.py

STEP 3 — Open the app:
    Open public/index.html with Live Server in VS Code
    OR run:  npm install && npm start

After conversion, public/model/tfjs_model/ will contain:
    model.json   — model architecture
    *.bin        — weight shards
"""

import os
import sys
import shutil

KERAS_FILE   = os.path.join("public", "model", "AgroMind.keras")
OUTPUT_DIR   = os.path.join("public", "model", "tfjs_model")

def main():
    print("🌱 AgroMind Model Converter")
    print("=" * 50)

    # Check keras file exists
    if not os.path.exists(KERAS_FILE):
        print(f"❌ Model file not found: {KERAS_FILE}")
        print("   Make sure AgroMind.keras is in public/model/")
        sys.exit(1)

    print(f"✅ Found: {KERAS_FILE}  ({os.path.getsize(KERAS_FILE)/1024/1024:.1f} MB)")

    # Check tensorflowjs
    try:
        import tensorflowjs as tfjs
        import tensorflow as tf
        print(f"✅ TensorFlow   {tf.__version__}")
        print(f"✅ TensorFlow.js converter ready")
    except ImportError as e:
        print(f"\n❌ Missing package: {e}")
        print("\nInstall with:")
        print("   pip install tensorflowjs tensorflow")
        sys.exit(1)

    # Clean output dir
    if os.path.exists(OUTPUT_DIR):
        shutil.rmtree(OUTPUT_DIR)
    os.makedirs(OUTPUT_DIR)

    print(f"\n⏳ Converting {KERAS_FILE} → {OUTPUT_DIR} …")
    print("   (this may take 30–60 seconds)\n")

    try:
        import tensorflow as tf
        import tensorflowjs as tfjs

        model = tf.keras.models.load_model(KERAS_FILE)
        print(f"   Model loaded: input={model.input_shape}  output={model.output_shape}")

        tfjs.converters.save_keras_model(model, OUTPUT_DIR)

        files = os.listdir(OUTPUT_DIR)
        total = sum(os.path.getsize(os.path.join(OUTPUT_DIR, f)) for f in files)
        print(f"\n✅ Conversion complete!")
        print(f"   Output: {OUTPUT_DIR}/")
        for f in sorted(files):
            sz = os.path.getsize(os.path.join(OUTPUT_DIR, f))
            print(f"     {f:40s}  {sz/1024:.0f} KB")
        print(f"   Total: {total/1024/1024:.1f} MB")
        print(f"\n🚀 Now open the app:")
        print(f"   npm install && npm start")
        print(f"   OR right-click public/index.html → Open with Live Server")

    except Exception as e:
        print(f"\n❌ Conversion failed: {e}")
        import traceback; traceback.print_exc()
        sys.exit(1)

if __name__ == "__main__":
    main()

File: convert_model.js
/**
 * ╔══════════════════════════════════════════════════════╗
 * ║  AgroMind — convert_model.js                        ║
 * ║  Converts AgroMind.keras → TF.js format             ║
 * ║  Uses Node.js — no Python/uvloop issues on Windows  ║
 * ╚══════════════════════════════════════════════════════╝
 *
 * STEP 1 — Install (run once, inside the agromind/ folder):
 *   npm install @tensorflow/tfjs-node
 *
 * STEP 2 — Run:
 *   node convert_model.js
 *
 * Output: public/model/tfjs_model/model.json + .bin shards
 */

const tf   = require('@tensorflow/tfjs-node');
const path = require('path');
const fs   = require('fs');

const KERAS_PATH = path.join(__dirname, 'public', 'model', 'AgroMind.keras');
const OUT_DIR    = path.join(__dirname, 'public', 'model', 'tfjs_model');

async function convert() {
  console.log('\n🌱 AgroMind Model Converter (Node.js)');
  console.log('='.repeat(50));

  // Check keras file exists
  if (!fs.existsSync(KERAS_PATH)) {
    console.error(`\n❌ Not found: ${KERAS_PATH}`);
    console.error('   Make sure AgroMind.keras is in public/model/');
    process.exit(1);
  }

  const sizeMB = (fs.statSync(KERAS_PATH).size / 1024 / 1024).toFixed(1);
  console.log(`✅ Found: AgroMind.keras (${sizeMB} MB)`);
  console.log('\n⏳ Loading model into TensorFlow.js…');
  console.log('   (first load may take 30–60 seconds)\n');

  try {
    // Load the .keras file
    const model = await tf.loadLayersModel(`file://${KERAS_PATH}`);

    console.log(`✅ Model loaded`);
    console.log(`   Input  shape: ${JSON.stringify(model.inputs[0].shape)}`);
    console.log(`   Output shape: ${JSON.stringify(model.outputs[0].shape)}`);
    console.log(`   Parameters:   ${model.countParams().toLocaleString()}`);

    // Clean output dir
    if (fs.existsSync(OUT_DIR)) fs.rmSync(OUT_DIR, { recursive: true });
    fs.mkdirSync(OUT_DIR, { recursive: true });

    console.log(`\n⏳ Saving to ${OUT_DIR} …`);

    // Save as TF.js format
    await model.save(`file://${OUT_DIR}`);

    // Show output files
    const files = fs.readdirSync(OUT_DIR);
    const total = files.reduce((s, f) =>
      s + fs.statSync(path.join(OUT_DIR, f)).size, 0);

    console.log('\n✅ Conversion complete!');
    console.log(`   Output: public/model/tfjs_model/`);
    files.sort().forEach(f => {
      const kb = (fs.statSync(path.join(OUT_DIR, f)).size / 1024).toFixed(0);
      console.log(`     ${f.padEnd(40)} ${kb} KB`);
    });
    console.log(`   Total: ${(total / 1024 / 1024).toFixed(1)} MB`);
    console.log('\n🚀 Done! Now start the app:');
    console.log('   npm start');
    console.log('   OR right-click public/index.html → Open with Live Server\n');

  } catch (err) {
    console.error('\n❌ Conversion failed:', err.message);
    console.error('\nFull error:');
    console.error(err);
    process.exit(1);
  }
}

convert();
